content/en/blog/2021-12-06-fake-it-when-you-make-it.md

```diff
@@ -0,0 +1,262 @@
+---
+title: "Fake It When You Make It: Creating Mocks in Go"
+slug: "fake-it-when-you-make-it"
+date: "2021-12-06T10:39:45-06:00"
+draft: true
+image_webp: images/blog/lake_reflection.webp
+image: images/blog/lake_reflection.jpg
+author: Patrick Deziel
+description: "A mock is a common pattern for unit testing interdependent services. In this post we will explore creating mocks in Go."
+---
+
+In the age of microservices and containerized applications, software is less monolithic and more interdependent. How do we write tests which account for this new reality? One common strategy is to create mocks for services. In this post, we'll explore some of the challenges with testing in Go and dive into some concrete examples where creating mocks can help alleviate these problems.
+<!--more-->
+
+## Testing An Existing Codebase
+
+As developers we'd like to think that we're following test-driven development or at least "development with tests". However, sometimes things need to be built quickly or we find ourselves inheriting an existing codebase. How do we write tests without fundamentally changing the source code?
+
+At Rotational we are interested in building intelligent distributed systems. This adds an additional layer to our testing problem because such projects often involve many external dependencies. A common strategy for dealing with these dependencies is to use mocks.
```

Comment:
It might be interesting to briefly introduce a specific example here between the sentences "...involve many external dependencies." and "A common strategy..."

Since we reference the SendGrid example later in the post. perhaps we could cue that up here? It might be a single brief sentence accompanied by a rough schema diagram, e.g.:

 "...involve many external dependencies. Consider for example an application the enables decentralized exchange of cryptographic information, which might require external calls to email delivery services, certificate authorities, and authentication APIs as well as internal calls to token managers, such as the one in the below sketch. A common strategy..."

Reply:
That's a good suggestion, I added a sketch; let me know what you think.